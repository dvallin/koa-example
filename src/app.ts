import { start } from './server'
import { production } from './mode'

export const server = start(production())
