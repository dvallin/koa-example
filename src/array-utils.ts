function isDefined<T>(t: T | undefined): boolean {
  return t !== undefined
}

export function pluck<T, K extends keyof T>(values: T[], key: K): T[K][] {
  return values.map(v => v[key]).filter(isDefined)
}
