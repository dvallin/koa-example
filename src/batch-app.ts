import { runMode } from './framework/batch/runner'
import { mergeExports } from './framework'
import { ConsoleLogger } from './framework/logger'
import { wrapBatchHandler } from './framework/batch/handlers'

runMode({
  loggerProvider: (name: string) => new ConsoleLogger(name),
  exports: mergeExports({
    handlers: [wrapBatchHandler(() => Promise.resolve())],
  }),
})
