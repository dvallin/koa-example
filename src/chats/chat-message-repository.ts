import { Postgres, createQuery, createPreparedQuery } from '../io/postgres'
import { QueryConfig } from 'pg'
import { Message } from '.'

export interface ChatMessageRepository {
  create(message: Message): Promise<void>
}

const TABLE_NAME = 'chat_messages'

export const migrations: QueryConfig[] = [
  `CREATE TABLE IF NOT EXISTS ${TABLE_NAME} (
        room  VARCHAR(45),
        date  TIMESTAMP,
        message   TEXT,
        PRIMARY KEY(room, date)
      );`,
].map(text => createQuery(text))

const commands = {
  'create-message': `INSERT INTO ${TABLE_NAME} (room, date, message) VALUES ($1, $2, $3);`,
}

export class ChatMessageRepositoryImpl implements ChatMessageRepository {
  constructor(private readonly postgres: Postgres) {}

  async create(message: Message): Promise<void> {
    const command = createPreparedQuery(commands, 'create-message', message.room, message.date, message.message)
    await this.postgres.performQuery(command)
  }
}
