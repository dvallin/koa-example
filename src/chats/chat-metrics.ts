import { Metrics, Gauge } from '../framework/metrics'

export class ChatMetrics {
  public readonly openSockets: Gauge

  constructor(metrics: Metrics) {
    this.openSockets = metrics.gauge('openSockets', 'open sockets')
  }
}
