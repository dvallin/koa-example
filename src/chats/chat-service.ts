import { ChatMessageRepository } from './chat-message-repository'
import { Message } from '.'

export interface ChatService {
  received(message: Message): Promise<void>
}

export class ChatServiceImpl implements ChatService {
  constructor(private readonly messageRepository: ChatMessageRepository) {}

  received(message: Message): Promise<void> {
    return this.messageRepository.create(message)
  }
}
