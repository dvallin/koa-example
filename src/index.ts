import * as Users from './users'
import * as Chats from './chats'
import * as IO from './io'
import { pluck } from './array-utils'
import { QueryConfig } from 'pg'
import { Middleware } from 'koa'
import { Socket, Server } from 'socket.io'

export type Hook = () => Promise<void>
export type SocketHandler = (server: Server, socket: Socket) => void

export interface ModuleExports {
  startup: Hook
  shutdown: Hook

  middlewares: Middleware[]
  migrations: QueryConfig[]
  socketHandlers: SocketHandler[]
}

export interface Module<T> {
  components: T
  exports: Partial<ModuleExports>
}

export interface Mode {
  io: IO.Components
  users: Users.Components
  chats: Chats.Components

  exports: ModuleExports
}

async function callHooks(hooks: Hook[]): Promise<void> {
  await Promise.all(hooks.map(hook => hook()))
}

export function mergeExports(...moduleExports: Partial<ModuleExports>[]): ModuleExports {
  const startup = (): Promise<void> => callHooks(pluck(moduleExports, 'startup'))
  const shutdown = (): Promise<void> => callHooks(pluck(moduleExports, 'shutdown'))
  const migrations = [].concat(...pluck(moduleExports, 'migrations'))
  const middlewares = [].concat(...pluck(moduleExports, 'middlewares'))
  const socketHandlers = [].concat(...pluck(moduleExports, 'socketHandlers'))
  return { startup, shutdown, migrations, middlewares, socketHandlers }
}

export function production(): Mode {
  const io = IO.production()
  const users = Users.production(io.components)
  const chats = Chats.production(io.components)
  return {
    io: io.components,
    users: users.components,
    chats: chats.components,
    exports: mergeExports(io.exports, users.exports, chats.exports),
  }
}
