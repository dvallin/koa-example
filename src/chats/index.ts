import * as IO from '../io'
import { Module } from '..'
import { buildChatSocketHandler } from './chat-socket-handler'
import { ChatServiceImpl, ChatService } from './chat-service'
import { ChatMessageRepositoryImpl, ChatMessageRepository, migrations } from './chat-message-repository'

export interface Components {
  messageRepository: ChatMessageRepository
  service: ChatService
}

export interface Message {
  room: string
  date: Date
  message: string
}

export function production(io: IO.Components): Module<Components> {
  const messageRepository = new ChatMessageRepositoryImpl(io.postgres)
  const service = new ChatServiceImpl(messageRepository)
  const chatSocketHandler = buildChatSocketHandler(service)
  return {
    components: { messageRepository, service },
    exports: { socketHandlers: [chatSocketHandler], migrations: migrations },
  }
}
