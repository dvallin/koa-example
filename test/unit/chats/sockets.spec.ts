import * as ws from 'socket.io-client'
import { testMode } from '../../test-wrappers'
import { Message } from '../../../src/chats'

import { testing } from '../../test-modes'
import { MockPostgres } from '../../test-modes/mock-database'

interface SerializedMessage {
  room: string
  message: string
  date: string
}

describe('chat socket', () => {
  testMode(testing, 'can join room and send messages to it', async (_server, mode, port) => {
    return new Promise(resolve => {
      const postgres = mode.io.postgres as MockPostgres
      const socket1Messages: Message[] = []
      const socket2Messages: Message[] = []

      const tryResolve = () => {
        // both clients have received 2 messages
        if (socket1Messages.length == 2 && socket2Messages.length == 2) {
          const messages: Message[] = postgres.performedQueries
            .filter(q => q.name === 'create-message')
            .map(q => ({ room: q.values[0], date: q.values[1], message: q.values[2] }))

          // all messages are saved and every client has received them in order
          expect(messages.map(m => ({ ...m, date: '<date overriden>' }))).toMatchSnapshot()
          expect(socket1Messages).toEqual(messages)
          expect(socket2Messages).toEqual(messages)
          resolve()
        }
      }

      const socket1 = ws(`http://localhost:${port}`)
      const socket2 = ws(`http://localhost:${port}`)
      socket1.emit('join', 'secret room')
      socket2.emit('join', 'secret room')

      socket1.emit('chat message', 'socket1> hello socket 2')
      socket2.emit('chat message', 'socket2> hello socket 1')

      const receiveMessage = (into: Message[]) => {
        return (msg: SerializedMessage) => {
          into.push({ ...msg, date: new Date(msg.date) })
          tryResolve()
        }
      }

      socket1.on('chat message', receiveMessage(socket1Messages))
      socket2.on('chat message', receiveMessage(socket2Messages))
    })
  })
})
