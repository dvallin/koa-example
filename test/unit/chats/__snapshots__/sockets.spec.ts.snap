// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat socket with running server can join room and send messages to it 1`] = `
Array [
  Object {
    "date": "<date overriden>",
    "message": "socket1> hello socket 2",
    "room": "secret room",
  },
  Object {
    "date": "<date overriden>",
    "message": "socket2> hello socket 1",
    "room": "secret room",
  },
]
`;
